File: internal/manager/agent/sidecar.go
package agent

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"unicode/utf8"

	"github.com/codeasier/mtls-router/internal/manager/agent/modelconfig"
)

const maxSidecarSize = 2 << 20

type lastAppliedState struct {
	Version       int                       `json:"version"`
	KeyGeneration string                    `json:"key_generation"`
	Agents        map[Kind]lastAppliedAgent `json:"agents"`
}

type lastAppliedAgent struct {
	ModelConfig json.RawMessage   `json:"model_config"`
	Files       []lastAppliedFile `json:"files"`
	OwnedPaths  []string          `json:"owned_paths"`
}

type lastAppliedFile struct {
	Role        string `json:"role"`
	Path        string `json:"path"`
	RevisionMAC string `json:"revision_mac"`
}

func (s *Service) sidecarPath() string { return filepath.Join(s.stateDir, sidecarFileName) }

func (s *Service) readSidecar() (lastAppliedState, fileRevision, []byte, os.FileMode, error) {
	revision, content, mode, err := s.readKeyedRevision(s.sidecarPath(), revisionContextSidecar)
	if err != nil {
		return lastAppliedState{}, fileRevision{}, nil, 0, operationError(CodeModelStateInvalid, "Agent model state cannot be read")
	}
	if !revision.Exists {
		return lastAppliedState{Version: 1, KeyGeneration: s.keyGeneration, Agents: map[Kind]lastAppliedAgent{}}, revision, nil, 0o600, nil
	}
	info, err := os.Stat(s.sidecarPath())
	if err != nil || !privatePermissionsOK(s.sidecarPath(), false, info.Mode()) || len(content) == 0 || len(content) > maxSidecarSize {
		return lastAppliedState{}, fileRevision{}, nil, 0, operationError(CodeModelStateInvalid, "Agent model state is invalid")
	}
	var state lastAppliedState
	if err := strictJSON(content, &state); err != nil || validateSidecarState(state, s.keyGeneration) != nil {
		return lastAppliedState{}, fileRevision{}, nil, 0, operationError(CodeModelStateInvalid, "Agent model state is invalid")
	}
	canonical, err := modelconfig.CanonicalValue(state)
	if err != nil || !bytes.Equal(content, canonical) {
		return lastAppliedState{}, fileRevision{}, nil, 0, operationError(CodeModelStateInvalid, "Agent model state is not canonical")
	}
	return state, revision, content, mode, nil
}

func validateSidecarState(state lastAppliedState, generation string) error {
	if state.Version != 1 || state.KeyGeneration != generation || state.Agents == nil || len(state.Agents) > 3 {
		return errors.New("invalid state header")
	}
	for kind, section := range state.Agents {
		if kind != ClaudeCode && kind != OpenCode && kind != Codex || len(section.ModelConfig) == 0 || len(section.Files) == 0 {
			return errors.New("invalid Agent state")
		}
		seenRoles := map[string]bool{}
		for _, file := range section.Files {
			if (file.Role != "config" && file.Role != "auth") || seenRoles[file.Role] || !filepath.IsAbs(file.Path) || file.RevisionMAC == "" || len(file.RevisionMAC) > 128 {
				return errors.New("invalid state file")
			}
			seenRoles[file.Role] = true
		}
		if kind != Codex && len(section.Files) != 1 || kind == Codex && (len(section.Files) != 2 || !seenRoles["config"] || !seenRoles["auth"]) {
			return errors.New("inconsistent state files")
		}
		if !sort.StringsAreSorted(section.OwnedPaths) {
			return errors.New("noncanonical owned paths")
		}
		for i, path := range section.OwnedPaths {
			if path == "" || len(path) > 4096 || strings.TrimSpace(path) != path || i > 0 && path == section.OwnedPaths[i-1] {
				return errors.New("invalid owned path")
			}
		}
	}
	return nil
}

func strictJSON(content []byte, dst any) error {
	if !utf8.Valid(content) || hasDuplicateJSONKey(content) {
		return errors.New("invalid JSON")
	}
	decoder := json.NewDecoder(bytes.NewReader(content))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(dst); err != nil {
		return err
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		return errors.New("trailing JSON")
	}
	return nil
}

func hasDuplicateJSONKey(content []byte) bool {
	decoder := json.NewDecoder(bytes.NewReader(content))
	var walk func(bool) bool
	walk = func(object bool) bool {
		if object {
			seen := map[string]bool{}
			for decoder.More() {
				token, err := decoder.Token()
				key, ok := token.(string)
				if err != nil || !ok || seen[key] {
					return true
				}
				seen[key] = true
				if walk(false) {
					return true
				}
			}
			_, err := decoder.Token()
			return err != nil
		}
		token, err := decoder.Token()
		if err != nil {
			return true
		}
		if delimiter, ok := token.(json.Delim); ok {
			switch delimiter {
			case '{':
				return walk(true)
			case '[':
				for decoder.More() {
					if walk(false) {
						return true
					}
				}
				_, err := decoder.Token()
				return err != nil
			}
		}
		return false
	}
	return walk(false)
}

func sidecarSection(config *modelconfig.Config, kind Kind) (json.RawMessage, []string, error) {
	var section any
	var owned []string
	switch kind {
	case ClaudeCode:
		section = config.Claude
		owned = make([]string, 0, len(claudeFixedEnvKeys)+len(config.Claude.Extra))
		for _, key := range claudeFixedEnvKeys {
			owned = append(owned, "env."+key)
		}
		for key := range config.Claude.Extra {
			owned = append(owned, "env."+key)
		}
	case OpenCode:
		section = config.OpenCode
		owned = []string{"model", "provider.mtls-router"}
	case Codex:
		section = config.Codex
		owned = append([]string{"model_providers.mtls-router", "auth.auth_mode", "auth.OPENAI_API_KEY"}, codexManagedRootKeys...)
	}
	sort.Strings(owned)
	raw, err := modelconfig.CanonicalValue(section)
	return raw, owned, err
}

func (s *Service) appendSidecarPlan(plan writePlan, key []byte) (writePlan, error) {
	state := plan.sidecar
	if state.Agents == nil {
		state.Agents = map[Kind]lastAppliedAgent{}
	}
	for _, kind := range plan.selected {
		section, owned, err := sidecarSection(plan.input.config, kind)
		if err != nil {
			return writePlan{}, operationError(CodeModelStateInvalid, "Agent model state could not be created")
		}
		entry := lastAppliedAgent{ModelConfig: section, OwnedPaths: owned}
		for i := range plan.files {
			file := &plan.files[i]
			if file.agent != kind || file.scope != scopeAgent {
				continue
			}
			output, err := file.render(key)
			if err != nil {
				return writePlan{}, operationError(CodeWriteFailed, "could not render an Agent configuration file")
			}
			mode := file.targetMode
			if !file.targetRevision.Exists {
				mode = 0o600
			}
			revision, err := s.keyedRevisionForContent(output, mode, revisionContextAgentFile, file.targetPath)
			zeroBytes(output)
			if err != nil {
				return writePlan{}, operationError(CodeModelStateInvalid, "Agent model state could not be created")
			}
			entry.Files = append(entry.Files, lastAppliedFile{Role: file.role, Path: file.targetPath, RevisionMAC: revisionTokenValue(revision)})
		}
		state.Agents[kind] = entry
	}
	state.Version = 1
	state.KeyGeneration = s.keyGeneration
	content, err := modelconfig.CanonicalValue(state)
	if err != nil || len(content) > maxSidecarSize {
		return writePlan{}, operationError(CodeModelStateInvalid, "Agent model state could not be created")
	}
	mode := plan.sidecarMode
	if !plan.sidecarRevision.Exists {
		mode = 0o600
	}
	contentCopy := append([]byte(nil), content...)
	operation := OperationCreate
	if plan.sidecarRevision.Exists {
		operation = OperationReplace
	}
	plan.files = append(plan.files, plannedFile{
		scope: scopeManagerState, role: "state", format: FormatJSON,
		sourcePath: s.sidecarPath(), targetPath: s.sidecarPath(), operation: operation,
		sourceRevision: plan.sidecarRevision, targetRevision: plan.sidecarRevision,
		sourceContent: plan.sidecarContent, sourceMode: plan.sidecarMode, targetMode: mode,
		backupRequired: plan.sidecarRevision.Exists, backupSource: s.sidecarPath(), restoreFrom: s.sidecarPath(),
		render: func([]byte) ([]byte, error) { return append([]byte(nil), contentCopy...), nil },
	})
	return plan, nil
}

File: desktop/src-tauri/src/manager_core/agent/sidecar.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use super::detect::strip_jsonc;
use super::files::private_permissions_ok;
use super::modelconfig::{
    canonical_value, config_to_value, decode_value, Agent, Config, JsonNumber, JsonValue,
    TokenSigner,
};
use super::plan::{
    clean_path, ConfigMode, FileRender, FileRevision, JournalScope, Operation, PlannedFile,
    WritePlan,
};
use super::render::{claude_owned_env_keys, codex_managed_config_keys};
use super::revision::{keyed_revision_for_content, read_keyed_revision};
use super::trust::SIDECAR_FILE_NAME;
use super::types::OperationError;
use crate::protocol::ErrorCode;

pub const MAX_SIDECAR_SIZE: usize = 2 << 20;

#[derive(Clone, Debug, Default, PartialEq)]
pub struct LastAppliedState {
    pub version: i64,
    pub key_generation: String,
    pub agents: HashMap<Agent, LastAppliedAgent>,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct LastAppliedAgent {
    pub model_config: JsonValue,
    pub files: Vec<LastAppliedFile>,
    pub owned_paths: Vec<String>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct LastAppliedFile {
    pub role: String,
    pub path: String,
    pub revision_mac: String,
}

pub fn sidecar_path(state_dir: &Path) -> PathBuf {
    state_dir.join(SIDECAR_FILE_NAME)
}

pub fn empty_sidecar(generation: &str) -> LastAppliedState {
    LastAppliedState {
        version: 1,
        key_generation: generation.to_owned(),
        agents: HashMap::new(),
    }
}

pub fn read_sidecar(
    signer: &TokenSigner,
    state_dir: &Path,
    generation: &str,
) -> Result<(LastAppliedState, FileRevision, Vec<u8>, u32), OperationError> {
    let path = sidecar_path(state_dir);
    let (revision, content, mode) =
        read_keyed_revision(signer, &path, super::plan::REVISION_CONTEXT_SIDECAR).map_err(
            |_| {
                OperationError::new(
                    ErrorCode::ModelStateInvalid,
                    "Agent model state cannot be read",
                )
            },
        )?;
    if !revision.exists {
        return Ok((empty_sidecar(generation), revision, Vec::new(), 0o600));
    }
    let info = std::fs::metadata(&path).map_err(|_| sidecar_invalid())?;
    if !private_permissions_ok(&path, false, unix_mode(&info))
        || content.is_empty()
        || content.len() > MAX_SIDECAR_SIZE
    {
        return Err(sidecar_invalid());
    }
    let value = decode_value(&content).map_err(|_| sidecar_invalid())?;
    let state = parse_sidecar_state(&value)?;
    validate_sidecar_state(&state, generation)?;
    let canonical = sidecar_to_canonical(&state)?;
    if content != canonical {
        return Err(OperationError::new(
            ErrorCode::ModelStateInvalid,
            "Agent model state is not canonical",
        ));
    }
    Ok((state, revision, content, mode))
}

pub fn validate_sidecar_state(
    state: &LastAppliedState,
    generation: &str,
) -> Result<(), OperationError> {
    if state.version != 1 || state.key_generation != generation || state.agents.len() > 3 {
        return Err(sidecar_invalid());
    }
    for (kind, section) in &state.agents {
        if matches!(section.model_config, JsonValue::Null) || section.files.is_empty() {
            return Err(sidecar_invalid());
        }
        let mut seen = HashMap::new();
        for file in &section.files {
            if (file.role != "config" && file.role != "auth")
                || seen.contains_key(&file.role)
                || !Path::new(&file.path).is_absolute()
                || file.revision_mac.is_empty()
                || file.revision_mac.len() > 128
            {
                return Err(sidecar_invalid());
            }
            seen.insert(file.role.clone(), true);
        }
        let files_ok = match kind {
            Agent::Claude | Agent::OpenCode => section.files.len() == 1,
            Agent::Codex => {
                section.files.len() == 2 && seen.contains_key("config") && seen.contains_key("auth")
            }
        };
        if !files_ok {
            return Err(sidecar_invalid());
        }
        if !section.owned_paths.windows(2).all(|pair| pair[0] < pair[1])
            && section.owned_paths.len() > 1
        {
            return Err(sidecar_invalid());
        }
        for (index, path) in section.owned_paths.iter().enumerate() {
            if path.is_empty()
                || path.len() > 4096
                || path.trim() != path
                || (index > 0 && path == &section.owned_paths[index - 1])
            {
                return Err(sidecar_invalid());
            }
        }
    }
    Ok(())
}

pub fn sidecar_to_canonical(state: &LastAppliedState) -> Result<Vec<u8>, OperationError> {
    canonical_value(&sidecar_to_value(state)?).map_err(|_| sidecar_invalid())
}

pub fn append_sidecar_plan(
    signer: &TokenSigner,
    state_dir: &Path,
    generation: &str,
    mut plan: WritePlan,
    key: &str,
) -> Result<WritePlan, OperationError> {
    let mut state = plan.sidecar.clone();
    for kind in &plan.selected {
        let previous = state.agents.get(kind).cloned().unwrap_or_default();
        let (model_config, owned) = sidecar_section(
            &plan.input.config,
            *kind,
            &plan.files,
            plan.input.own_root_model,
            &previous,
        )?;
        let mut entry = LastAppliedAgent {
            model_config,
            files: Vec::new(),
            owned_paths: owned,
        };
        for file in plan
            .files
            .iter()
            .filter(|file| file.agent == Some(*kind) && file.scope == JournalScope::Agent)
        {
            let mut output = file.output(&plan.input, key).map_err(|_| {
                OperationError::new(
                    ErrorCode::WriteFailed,
                    "could not render an Agent configuration file",
                )
            })?;
            let mode = if file.target_revision.exists {
                file.target_mode
            } else {
                0o600
            };
            let revision = keyed_revision_for_content(
                signer,
                &output,
                mode,
                super::plan::REVISION_CONTEXT_AGENT_FILE,
                &file.target_path,
            )
            .map_err(|_| sidecar_create_failed())?;
            super::plan::zero_bytes(&mut output);
            entry.files.push(LastAppliedFile {
                role: file.role.clone(),
                path: file.target_path.clone(),
                revision_mac: revision.token_value(),
            });
        }
        state.agents.insert(*kind, entry);
    }
    state.version = 1;
    state.key_generation = generation.to_owned();
    let content = sidecar_to_canonical(&state)?;
    if content.len() > MAX_SIDECAR_SIZE {
        return Err(sidecar_create_failed());
    }
    let mode = if plan.sidecar_revision.exists {
        plan.sidecar_mode
    } else {
        0o600
    };
    let operation = if plan.sidecar_revision.exists {
        Operation::Replace
    } else {
        Operation::Create
    };
    let path = sidecar_path(state_dir).to_string_lossy().into_owned();
    plan.sidecar = state;
    plan.files.push(PlannedFile {
        agent: None,
        mode: ConfigMode::Merge,
        format: super::types::Format::Json,
        source_path: path.clone(),
        target_path: path.clone(),
        operation,
        contains_api_key: false,
        preserves: Vec::new(),
        warning: String::new(),
        source_revision: plan.sidecar_revision.clone(),
        target_revision: plan.sidecar_revision.clone(),
        source_content: plan.sidecar_content.clone(),
        source_mode: plan.sidecar_mode,
        target_mode: mode,
        backup_required: plan.sidecar_revision.exists,
        backup_source: path.clone(),
        backup_path: String::new(),
        restore_from: path,
        render: Some(FileRender::Static(content)),
        role: "state".into(),
        companion_exists: false,
        syntax_invalid: false,
        scope: JournalScope::ManagerState,
    });
    Ok(plan)
}

fn sidecar_section(
    config: &Config,
    kind: Agent,
    files: &[PlannedFile],
    own_root_model: bool,
    previous: &LastAppliedAgent,
) -> Result<(JsonValue, Vec<String>), OperationError> {
    let encoded = config_to_value(config).map_err(|_| sidecar_create_failed())?;
    let object = encoded.as_object().ok_or_else(sidecar_create_failed)?;
    let section = object
        .get(kind.as_str())
        .cloned()
        .ok_or_else(sidecar_create_failed)?;
    let mut owned = match kind {
        Agent::Claude => config
            .claude
            .as_ref()
            .map(claude_owned_env_keys)
            .unwrap_or_default()
            .into_iter()
            .map(|key| format!("env.{key}"))
            .collect(),
        Agent::OpenCode => {
            let mut paths = vec!["provider.mtls-router".to_owned()];
            if open_code_plan_owns_root_model(files, own_root_model, previous) {
                paths.push("model".into());
            }
            paths
        }
        Agent::Codex => {
            let mut paths = vec![
                "model_providers.mtls-router".to_owned(),
                "auth.auth_mode".to_owned(),
                "auth.OPENAI_API_KEY".to_owned(),
            ];
            if let Some(codex) = &config.codex {
                for key in codex_managed_config_keys(codex, "") {
                    if key != "model_providers" {
                        paths.push(key);
                    }
                }
            }
            paths
        }
    };
    owned.sort();
    owned.dedup();
    Ok((section, owned))
}

fn open_code_plan_owns_root_model(
    files: &[PlannedFile],
    own_root_model: bool,
    previous: &LastAppliedAgent,
) -> bool {
    for file in files {
        if file.agent != Some(Agent::OpenCode) || file.scope != JournalScope::Agent {
            continue;
        }
        if file.mode == ConfigMode::Rebuild || own_root_model || !file.source_revision.exists {
            return true;
        }
        let mut content = file.source_content.clone();
        if Path::new(&file.source_path)
            .extension()
            .and_then(|value| value.to_str())
            == Some("jsonc")
        {
            match strip_jsonc(&content) {
                Ok(stripped) => content = stripped,
                Err(()) => return false,
            }
        }
        if let Some(root) = super::detect::decode_object(&content) {
            let Some(model) = root.get("model") else {
                return true;
            };
            return model
                .as_str()
                .is_some_and(|value| value.starts_with("mtls-router/"))
                && previous_owns_opencode_model(previous, file);
        }
    }
    false
}

fn previous_owns_opencode_model(previous: &LastAppliedAgent, file: &PlannedFile) -> bool {
    if !previous.owned_paths.iter().any(|path| path == "model") {
        return false;
    }
    previous.files.iter().any(|recorded| {
        recorded.role == "config"
            && file.role == "config"
            && clean_path(&recorded.path) == clean_path(&file.source_path)
    })
}

fn sidecar_to_value(state: &LastAppliedState) -> Result<JsonValue, OperationError> {
    let mut agents = HashMap::new();
    for (kind, section) in &state.agents {
        let files = JsonValue::Array(
            section
                .files
                .iter()
                .map(|file| {
                    JsonValue::object([
                        ("role".into(), JsonValue::String(file.role.clone())),
                        ("path".into(), JsonValue::String(file.path.clone())),
                        (
                            "revision_mac".into(),
                            JsonValue::String(file.revision_mac.clone()),
                        ),
                    ])
                })
                .collect(),
        );
        agents.insert(
            kind.as_str().to_owned(),
            JsonValue::object([
                ("model_config".into(), section.model_config.clone()),
                ("files".into(), files),
                (
                    "owned_paths".into(),
                    JsonValue::Array(
                        section
                            .owned_paths
                            .iter()
                            .map(|path| JsonValue::String(path.clone()))
                            .collect(),
                    ),
                ),
            ]),
        );
    }
    Ok(JsonValue::object([
        (
            "version".into(),
            JsonValue::Number(JsonNumber(state.version.to_string())),
        ),
        (
            "key_generation".into(),
            JsonValue::String(state.key_generation.clone()),
        ),
        ("agents".into(), JsonValue::Object(agents)),
    ]))
}

fn parse_sidecar_state(value: &JsonValue) -> Result<LastAppliedState, OperationError> {
    let object = value.as_object().ok_or_else(sidecar_invalid)?;
    require_keys(object, &["version", "key_generation", "agents"])?;
    let version = object
        .get("version")
        .and_then(JsonValue::as_number)
        .and_then(|value| value.as_str().parse().ok())
        .ok_or_else(sidecar_invalid)?;
    let key_generation = object
        .get("key_generation")
        .and_then(JsonValue::as_str)
        .ok_or_else(sidecar_invalid)?
        .to_owned();
    let agents_value = object.get("agents").ok_or_else(sidecar_invalid)?;
    let agents_object = agents_value.as_object().ok_or_else(sidecar_invalid)?;
    let mut agents = HashMap::new();
    for (key, value) in agents_object {
        let kind = Agent::parse(key).ok_or_else(sidecar_invalid)?;
        agents.insert(kind, parse_agent_section(value)?);
    }
    Ok(LastAppliedState {
        version,
        key_generation,
        agents,
    })
}

fn parse_agent_section(value: &JsonValue) -> Result<LastAppliedAgent, OperationError> {
    let object = value.as_object().ok_or_else(sidecar_invalid)?;
    require_keys(object, &["model_config", "files", "owned_paths"])?;
    let files = object
        .get("files")
        .and_then(JsonValue::as_array)
        .ok_or_else(sidecar_invalid)?;
    let owned = object
        .get("owned_paths")
        .and_then(JsonValue::as_array)
        .ok_or_else(sidecar_invalid)?;
    Ok(LastAppliedAgent {
        model_config: object
            .get("model_config")
            .cloned()
            .ok_or_else(sidecar_invalid)?,
        files: files.iter().map(parse_file).collect::<Result<_, _>>()?,
        owned_paths: owned
            .iter()
            .map(|value| {
                value
                    .as_str()
                    .map(ToOwned::to_owned)
                    .ok_or_else(sidecar_invalid)
            })
            .collect::<Result<_, _>>()?,
    })
}

fn parse_file(value: &JsonValue) -> Result<LastAppliedFile, OperationError> {
    let object = value.as_object().ok_or_else(sidecar_invalid)?;
    require_keys(object, &["role", "path", "revision_mac"])?;
    Ok(LastAppliedFile {
        role: required_string(object, "role")?,
        path: required_string(object, "path")?,
        revision_mac: required_string(object, "revision_mac")?,
    })
}

fn require_keys(object: &HashMap<String, JsonValue>, keys: &[&str]) -> Result<(), OperationError> {
    if object.len() != keys.len() || keys.iter().any(|key| !object.contains_key(*key)) {
        return Err(sidecar_invalid());
    }
    Ok(())
}

fn required_string(
    object: &HashMap<String, JsonValue>,
    key: &str,
) -> Result<String, OperationError> {
    object
        .get(key)
        .and_then(JsonValue::as_str)
        .map(ToOwned::to_owned)
        .ok_or_else(sidecar_invalid)
}

fn unix_mode(info: &std::fs::Metadata) -> u32 {
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        info.permissions().mode()
    }
    #[cfg(not(unix))]
    {
        0
    }
}

fn sidecar_invalid() -> OperationError {
    OperationError::new(ErrorCode::ModelStateInvalid, "Agent model state is invalid")
}

fn sidecar_create_failed() -> OperationError {
    OperationError::new(
        ErrorCode::ModelStateInvalid,
        "Agent model state could not be created",
    )
}
